controllers/common/consts.go

```diff
@@ -64,6 +64,9 @@ const (
 
 	// DBBackupCleanupFinalizer is the finalizer for cleaning up DatabaseClusterBackup.
 	DBBackupCleanupFinalizer = "everest.percona.com/dbb-cleanup"
+
+	// DBDeleteFinalizer is the finalizer for deleting the DatabaseCluster related resources.
+	DBDeleteFinalizer = "everest.percona.com/db-delete"
```

Comment:
Do we really need this finalizer? It seems to me that the only purpose it serves it to ensure that we see a 'deleting' state at least once.

 Currently the DB deletion may be blocked only due to the `dbb-delete` finalizer.

I think we can show a deleting state even without the use of this finalizer:

```diff
diff --git a/controllers/databasecluster_controller.go b/controllers/databasecluster_controller.go
index fb0313b..ff3927e 100644
--- a/controllers/databasecluster_controller.go
+++ b/controllers/databasecluster_controller.go
@@ -133,7 +133,8 @@ func (r *DatabaseClusterReconciler) reconcileDB(
                if done, err := p.Cleanup(ctx, db); err != nil {
                        return ctrl.Result{}, err
                } else if !done {
-                       return ctrl.Result{Requeue: true}, nil
+                       db.Status.Status = everestv1alpha1.AppStateDeleting
+                       return ctrl.Result{Requeue: true}, r.Status().Update(ctx, db)
                }
                return ctrl.Result{}, nil
        }
```

With this change, the outer controller simply checks if the inner Cleanup is complete. If not, we update the Status to deleting. We already have finalizers in place that will block the deletion until the clean-up of the respective child resources.

What do you think?

Reply:
This way, the inner Cleanup is really only responsible for the actual clean-up, and the outer controller updates the state to Deleting if the clean-up is not complete.

Since we already have finalizers that will prevent deletion, we don't need this new finalizer

Reply:
Or even better if we're doing the foreground deletion I mentioned in https://github.com/percona/everest-operator/pull/392#discussion_r1592129708 :

```diff
diff --git a/controllers/databasecluster_controller.go b/controllers/databasecluster_controller.go
index fb0313b..9f56419 100644
--- a/controllers/databasecluster_controller.go
+++ b/controllers/databasecluster_controller.go
@@ -130,12 +130,14 @@ func (r *DatabaseClusterReconciler) reconcileDB(
 ) (ctrl.Result, error) {
        // Handle any necessary cleanup.
        if !db.GetDeletionTimestamp().IsZero() {
-               if done, err := p.Cleanup(ctx, db); err != nil {
+               done, err := p.Cleanup(ctx, db)
+               if err != nil {
                        return ctrl.Result{}, err
-               } else if !done {
-                       return ctrl.Result{Requeue: true}, nil
                }
-               return ctrl.Result{}, nil
+               db.Status.Status = everestv1alpha1.AppStateDeleting
+               return ctrl.Result{
+                       Requeue: !done,
+               }, r.Status().Update(ctx, db)
        }

        // Set metadata.
```

Reply:
Thank you for the feedback, makes sense, let me rework the approach👍